Sources/SourceKitLSP/Swift/CodeActions/SyntaxRefactoringCodeActionProvider.swift

```diff
@@ -0,0 +1,75 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+import SwiftRefactor
+import SwiftSyntax
+
+/// Protocol that adapts a SyntaxRefactoringProvider (that comes from
+/// swift-syntax) into a SyntaxCodeActionProvider.
+protocol SyntaxRefactoringCodeActionProvider: SyntaxCodeActionProvider, SyntaxRefactoringProvider {
+  static var title: String { get }
+}
+
+/// SyntaxCodeActionProviders with a \c Void context can automatically be
+/// adapted provide a code action based on their refactoring operation.
+extension SyntaxRefactoringCodeActionProvider where Self.Context == Void {
+  static func codeActions(in scope: SyntaxCodeActionScope) -> [CodeAction] {
+    guard
+      let token = scope.firstToken,
+      let node = token.parent?.as(Input.self)
```

Comment:
Should we traverse the parents until we find a node of type `Input`? That way you don’t need to start eg. a `MigrateToNewIfLetSyntax` refactoring on the `if` token but you could also start it from within the condition or the body (not sure if we want the latter though, maybe we should stop the traversal at some pre-defined nodes like `CodeBlockItemListSyntax` and `MemberBlockItemListSyntax` to prevent that).

Similarly, the `OpaqueParameterToGeneric` refactoring action can’t currently be invoked from one of the `some` types, which might be confusing.

Reply:
I'm wondering if this is something that should be handled by the `MigrateToNewIfLetSyntax`, `OpaqueParameterToGeneric`, etc., refactorings themselves, because we don't necessarily know how far out in the tree we should go.

Reply:
That seems reasonable.